site-src/geps/gep-917.md

```diff
@@ -0,0 +1,162 @@
+like Gateway itself, are required for all implementations, but implementations
+may choose what Route objects they support. There will probably be some common
+sets of Route objects supported across similar implementations, but this proposal
+expects that we will make calls about what to call that common experience at a
+later date.
+- Support for fields or features that have Core, Extended, and
+ImplementationSpecific support. In particular, it must be possible for
+implementations to only support some subset of Extended fields, and to be able
+to use the framework for their own ImplementationSpecific features if required.
+- A testing suite that can validate that an implementation meets the conformance
+profiles it claims.
+- A way to retrieve conformance information in a machine-parseable way.
+
+## Proposal
+
+### Conformance profiles principles
+
+#### Basics
+
+The Gatway API project defines conformance purely in terms of what objects an
+implementation supports.
+By "supports" in this context, we mean the functionality covered by Core support.
+Extended support fields are handled by a separate mechanism.
+
+All implementations must support all the Core functions on the following objects:
+- GatewayClass
+- Gateway
+- ReferencePolicy
+
+The following objects are optional to support, but have defined behavior if you
+do:
+- HTTPRoute
+- TLSRoute
+- TCPRoute
+- UDPRoute
+
+For all of these objects, we should aim to have the usual range of tests for
+both the happy and unhappy paths for various types of operations.
+
+The conformance is versioned - it tracks the required features for a specific
+version of the API, and must be included in and updated by a version bump in the
+bundle version of the Gateway API. (The _bundle version_ is the thing that we
+mark as a "release", that looks like `v0.4.0`, not `v1alpha2`). 
+
+This will enable an implementation to say that it supports version whatever of
+the Gateway API. This is again similar to upstream in that implementations need
+to submit conformance test results for each version of Kubernetes they support.
+
+Because the support is defined in terms of the objects that an implementation
+supports, the conformance is composable, and orthogonal for each object type.
+It's valid to only support TLSRoute and not HTTPRoute, for example, although
+it would be odd.
+
+#### Interaction with existing support levels
+
+Conformance definitions will ensure that an implementation can provide all the
+features currently marked as "Core" support in the API documentation.
+
+Fields marked "Extended" support will eventually have conformance tests that
+lock in the behavior of that feature, and there will be a mechanism for implementations
+to tell the testing framework what extended fields they support.
+
+#### Testing framework
+
+The Gateway API project will provide a set of tests and harness to run them, such
+that an implementation may point the test harness at a Gateway or set of Gateways
```

Comment:
I've actually thought of conformance targeting a GatewayClass and then creating/testing Gateways, would that be workable?

Reply:
Maybe?
In-cluster proxies will need something akin to an operator for that to work.
Kong is not considering Gateway lifecycle management in the first iteration.

Reply:
If Kong is deployed behind a Service, will it be sufficiently aware of that to set an IP in Gateway status? I believe that's how Istio's currently setting this status, but not sure, @howardjohn would know best.

Reply:
I think that just putting in here that the testing may be pointed at a GatewayClass or Gateway will cover us no matter what.

But I think that full dynamic Gateway creation may be a ways away for a few implementations at least, Kong and Contour among them. So we should have a way to set whether a new Gateway can be created or not as well.